rag/structured_rag/entity_graph.py

```
@@ -0,0 +1,766 @@
+        for r in self.relationships:
+            if r.source_id == character_id:
+                events.update(r.events)
+        # Get events where character is a participant
+        for r in self.relationships:
+            if r.target_id == character_id and r.relation == "participant":
+                # Extract display name from namespaced ID
+                event_name = r.source_id.split(":", 1)[1] if ":" in r.source_id else r.source_id
+                events.add(event_name)
+        return list(events)
+    
+    def get_relevant_events(self, character_names: list[str], group_names: list[str] = None, context: str = "") -> list[str]:
```

Ruff (0.15.9)

[warning] PEP 484 prohibits implicit `Optional`

Convert to `T | None`

(RUF013)